src/libguac/local-lock.c

```diff
@@ -0,0 +1,218 @@
+ * nibble of the value.
+ *
+ * @param value
+ *     The key value containing the flag.
+ *
+ * @return
+ *     The flag indicating which lock is held, if any.
+ */
+static uintptr_t get_lock_flag(uintptr_t value) {
+    return value & 0xF;
+}
+
+/**
+ * Extract and return the lock count from the provided key. This returned value
+ * is the difference between the number of lock and unlock requests made by the
+ * current thread. This count is always stored in the remaining value after the
+ * least-significant nibble where the flag is stored.
+ *
+ * @param value
+ *     The key value containing the count.
+ *
+ * @return
+ *     The difference between the number of lock and unlock requests made by
+ *     the current thread.
+ */
+static uintptr_t get_lock_count(uintptr_t value) {
+    return value >> 4;
+}
+
+/**
+ * Given a flag indicating if and how the current thread controls a lock, and
+ * a count of the depth of lock requests, return a value containing the flag
+ * in the least-significant nibble, and the count in the rest.
+ *
+ * @param flag
+ *     A flag indiciating which lock, if any, is held by the current thread.
+ *
+ * @param count
+ *     The depth of the lock attempt by the current thread, i.e. the number of
+ *     lock requests minus unlock requests.
+ *
+ * @return
+ *     A value containing the flag in the least-significant nibble, and the
+ *     count in the rest, cast to a void* for thread-local storage.
+ */
+static void* get_value_from_flag_and_count(
+        uintptr_t flag, uintptr_t count) {
+    return (void*) ((flag & 0xF) | count << 4);
+}
+
+void guac_acquire_write_lock(guac_local_lock* local_lock) {
+
+    uintptr_t key_value = (uintptr_t) pthread_getspecific(local_lock->key);
+    uintptr_t flag = get_lock_flag(key_value);
+    uintptr_t count = get_lock_count(key_value);
+
+    /* If the current thread already holds the write lock, increment the count */
+    if (flag == GUAC_LOCAL_LOCK_WRITE_LOCK) {
+        pthread_setspecific(local_lock->key, get_value_from_flag_and_count(
+                flag, count + 1));
+
+        /* This thread already has the lock */
+        return;
+    }
+
+    /*
+     * The read lock must be released before the write lock can be acquired.
+     * This is a little odd because it may mean that a function further down
+     * the stack may have requested a read lock, which will get upgraded to a
+     * write lock by another function without the caller knowing about it. This
+     * shouldn't cause any issues, however.
+     */
+    if (key_value == GUAC_LOCAL_LOCK_READ_LOCK)
+        pthread_rwlock_unlock(&(local_lock->lock));
+
+    /* Acquire the write lock */
+    pthread_rwlock_wrlock(&(local_lock->lock));
+
+    /* Mark that the current thread has the lock, and increment the count */
+    pthread_setspecific(local_lock->key, get_value_from_flag_and_count(
+            GUAC_LOCAL_LOCK_WRITE_LOCK, count + 1));
```

Comment:
Well, it'll roll over back to 0, which might rely in the lock being acquired twice (resulting in undefined behavior), or the lock getting released early, resulting in potential simultaneous access to pieces of code that should only have one thread working with them at a time.

Are you worried enough about this that you want to add checks in the code for this condition?

Reply:
Moderately, yes. It seems unlikely that the counter would ever reach this magnitude, but I think at least that we should have defined behavior for that case.

The `pthread_mutex_lock` function's behavior for this case is to return `EAGAIN`:

> <dl>
>     <dt><code>EAGAIN</code></dt>
>     <dd>The mutex could not be acquired because the maximum number of recursive locks for mutex has been exceeded.</dd>
> </dl>

We could do similar (provide some sort of error handling) and leave it up to the caller to deal with the error condition.

Reply:
Ok, I added some error codes. Are you ok with not checking for the errors in the calling libguac code? We're not checking to see whether acquisition attempts of normal lock types work currently, and I happen to know that the maximum number of levels of reentrancy that we'll currently reach is `2,` which is significantly less than `2^60`.

Reply:
I think it's OK to leave as-is, unless there is any locking within a loop ... though 1,152,921,504,606,846,976 would be a _lot_ of locks.

Reply:
Yeah, this is probably the least of your problems if you find yourself acquiring that many locks.